pkg/stages/steps_install.go

```diff
@@ -324,3 +334,626 @@ depend() {
+					Permissions: 0755,
+					Owner:       0,
+					Group:       0,
+					Content: `[Unit]
+Description=kairos interactive-installer
+After=multi-user.target
+[Service]
+## Dont mark it as running until it finishes
+Type=oneshot
+# input/output to tty as its interactive
+# otherwise it will be silent and with no input
+StandardInput=tty
+StandardOutput=tty
+LimitNOFILE=49152
+ExecStartPre=-/bin/sh -c "dmesg -D"
+TTYPath=/dev/tty1
+RemainAfterExit=yes
+# Stop systemd messages on tty
+ExecStartPre=-/usr/bin/kill -SIGRTMIN+21 1
+ExecStart=/usr/bin/kairos-agent interactive-install --shell
+# Start systemd messages on tty
+ExecStartPost=-/usr/bin/kill -SIGRTMIN+20 1
+TimeoutStopSec=10s
+# Restart if it fails, like user doing control+c
+Restart=on-failure
+[Install]
+WantedBy=multi-user.target`,
+				},
+			},
+		},
+	}...)
+
+	return data
+}
+
+// GetInstallKairosBinariesStage directly installs the kairos binaries from the remote location
+// TODO: Ideally this should be able to be done wiht a yip plugin
+// Something like InstallFromGithubRelease
+// In some distros, this needs to run after the packages are installed due to the ca-certificates package
+// being installed later, as we are using https
+func GetInstallKairosBinariesStage(sis values.System, l types.KairosLogger) error {
+	var fips string
+	targetDir := "/usr/bin"
+	arch := sis.Arch.String()
+
+	if config.DefaultConfig.Fips {
+		fips = "-fips"
+	}
+
+	// Download the kairos-agent binary
+	url := fmt.Sprintf("https://github.com/kairos-io/kairos-agent/releases/download/%s/kairos-agent-%s-Linux-%s%s.tar.gz", values.GetAgentVersion(), values.GetAgentVersion(), arch, fips)
+	binaryName := "kairos-agent"
+
+	err := downloadAndExtractBinaryInMemory(l, url, binaryName, targetDir)
+	if err != nil {
+		return err
+	}
+	// Download the immucore binary
+	url = fmt.Sprintf("https://github.com/kairos-io/immucore/releases/download/%s/immucore-%s-Linux-%s%s.tar.gz", values.GetImmucoreVersion(), values.GetImmucoreVersion(), arch, fips)
+	binaryName = "immucore"
+	err = downloadAndExtractBinaryInMemory(l, url, binaryName, targetDir)
+	if err != nil {
+		return err
+	}
+
+	// Download the immucore binary
+	url = fmt.Sprintf("https://github.com/kairos-io/kcrypt-challenger/releases/download/%s/kcrypt-discovery-challenger-%s-Linux-%s%s.tar.gz", values.GetKcryptChallengerVersion(), values.GetKcryptChallengerVersion(), arch, fips)
+	binaryName = "kcrypt-discovery-challenger"
+	targetDir = "/system/discovery/"
+	// if the directory does not exist, create it
+	if _, err := os.Stat(targetDir); os.IsNotExist(err) {
+		err = os.MkdirAll(targetDir, 0755)
+		if err != nil {
+			l.Logger.Error().Err(err).Str("dir", targetDir).Msg("Failed to create directory")
+			return err
+		}
+	}
+	err = downloadAndExtractBinaryInMemory(l, url, binaryName, targetDir)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func downloadAndExtractBinaryInMemory(l types.KairosLogger, url, binaryName, targetDir string) error {
```

Comment:
This function needs to validate the downloaded binaries with a checksum. Especially since users are going to user kairos-init in their own infrastructure to build Kairos images (we can define the checksums [along with the versions](https://github.com/kairos-io/kairos-init/pull/61/files#diff-292b66a7a8fd7d0d19fc58ee9d8902d0100b842e45c137963ad4caa22f3a5527R27-R32)).

Also, having hardcoded versions in the kairos-init binary makes it more than a build tool and more of a specific "release" of Kairos, similar to what each version of the framework was. I don't remember if we had other things already that gave kairos-init such a role but I'm pointing this out to make sure it's a delibarate decision.

Reply:
Ideally all this binaries would be overridable like the framework is, so we ship init with latest always and auto updated by renovate, but you can always set whatever version you want, kind of like the framework now. So init versions are not tied to Kairos versions, whatever those are. We mark which binaries to use when we tag a Kairos version. It should also give us a more easy way of backporting stuff.


Good point on the checksums, they should already be available on the releases for the binaries so it should be easy to check those!